image-rs/src/signature/mechanism/simple/xrss.rs

```diff
@@ -0,0 +1,167 @@
+    #[serde(default)]
+    signatures: Vec<RegistrySignature>,
+}
+
+#[derive(Deserialize, Debug, PartialEq, Eq, Serialize, Default)]
+pub struct RegistrySignature {
+    #[serde(default, rename = "schemaVersion")]
+    pub schema_version: u32,
+    #[serde(default)]
+    pub name: String,
+    #[serde(default, rename = "type")]
+    pub signature_type: String,
+    #[serde(default)]
+    pub content: String,
+}
+
+impl RegistryClient {
+    pub fn new() -> RegistryClient {
+        RegistryClient {
+            client: reqwest::Client::new(),
+        }
+    }
+
+    async fn get_oauth_token_for_registry(
+        &self,
+        image: &Image,
+        auth: &RegistryAuth,
+    ) -> Result<String> {
+        match auth {
+            RegistryAuth::Anonymous => {
+                bail!("Trying to get signature from a registry without providing auth information")
+            }
+            RegistryAuth::Basic(username, password) => {
+                let scope = format!("repository:{}:pull", image.reference.repository());
+                let res = self
+                    .client
+                    .get(format!(
+                        "https://{}/oauth/token",
+                        image.reference.registry()
+                    ))
+                    .query(&[
+                        ("account", username),
+                        ("scope", &scope),
+                        ("service", &"registry".to_string()),
+                    ])
+                    .basic_auth(username, Some(password))
+                    .send()
+                    .await
+                    .map_err(|e| anyhow!("Failed to fetch oauth token for registry: {:?}", e))?;
+
+                let oauth_token_response_body: GetOAuthTokenResponse =
+                    res.json::<GetOAuthTokenResponse>().await.map_err(|e| {
+                        anyhow!(
+                            "Unexpected response from fetching oauth token from registry: {:?}",
+                            e
+                        )
+                    })?;
+
+                Ok(oauth_token_response_body.token)
+            }
+        }
+    }
+
+    pub async fn get_signatures_from_registry(
+        &self,
+        image: &Image,
+        digest: &Digest,
+        auth: &RegistryAuth,
+    ) -> Result<Vec<Vec<u8>>> {
+        let mut sigs: Vec<Vec<u8>> = simple::new_sigs_vec().await;
+
+        let res = self
+            .client
+            .get(format!("https://{}/v2/", image.reference.registry()))
+            .send()
+            .await
+            .map_err(|e| {
+                anyhow!(
+                    "Failed to query extensions supported by registry v2 endpoint: {:?}",
+                    e
+                )
+            })?;
+
+        let registry_signature_support_header = res.headers().get("x-registry-supports-signatures");
+        if registry_signature_support_header.is_some()
+            && registry_signature_support_header.unwrap() == "1"
```

Comment:
I noticed that here `unwrap()` is used. Will this never cause panic? If so, I suggest that add `.expect("...")` to the end of line 120.

Reply:
The response returned by `https://<registry_name>/v2/` may not include the `x-registry-supports-signatures` header so `registry_signature_support_header` can be `None`. So if the registry doesn't support the signatures API extension I would expect the code to not panic and return an empty `Vec`.

E.g., quay.io vs. ICR:

```
curl -v https://quay.io/v2/ 2>&1 | grep x-registry-supports-signatures
curl -v https://uk.icr.io/v2/ 2>&1 | grep x-registry-supports-signatures
< x-registry-supports-signatures: 1
```